cpp/src/core/core_number.cu

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * ---------------------------------------------------------------------------*
+ * @brief Core Number implementation
+ *
+ * @file core_number.cu
+ * --------------------------------------------------------------------------*/
+
+#include <cugraph.h>
+#include "utilities/error_utils.h"
+#include <Hornet.hpp>
+#include <Static/CoreNumber/CoreNumber.cuh>
+#include <rmm_utils.h>
+#include <nvgraph_gdf.h>
+
+gdf_error core_number_impl(gdf_graph *graph,
+                          int *core_number) {
+  using HornetGraph = hornet::gpu::HornetStatic<int>;
+  using HornetInit  = hornet::HornetInit<int>;
+  using CoreNumber  = hornets_nest::CoreNumberStatic;
+  HornetInit init(graph->numberOfVertices, graph->adjList->indices->size,
+      static_cast<int*>(graph->adjList->offsets->data),
+      static_cast<int*>(graph->adjList->indices->data));
+  HornetGraph hnt(init, hornet::DeviceType::DEVICE);
+  CoreNumber cn(hnt, core_number);
+  cn.run();
+  return GDF_SUCCESS;
+}
+
+gdf_error gdf_core_number(gdf_graph *graph,
+                          gdf_column *core_number) {
+  GDF_REQUIRE(graph->adjList != nullptr || graph->edgeList != nullptr, GDF_INVALID_API_CALL);
+  gdf_error err = gdf_add_adj_list(graph);
+  if (err != GDF_SUCCESS)
+    return err;
+  GDF_REQUIRE(graph->adjList->offsets->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(graph->adjList->indices->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(core_number->dtype == GDF_INT32, GDF_UNSUPPORTED_DTYPE);
+  GDF_REQUIRE(core_number->size == graph->numberOfVertices, GDF_COLUMN_SIZE_MISMATCH);
+
+  return core_number_impl(graph, static_cast<int*>(core_number->data));
+}
+
+struct LessThan {
+  int k;
+
+  LessThan(int _k) : k(_k) {}
+
+  __host__ __device__
+    bool operator()(const int core) {
+      return core < k;
+    }
+};
+
+gdf_error gdf_k_core(gdf_graph *in_graph,
```

Comment:
This function is doing more than just calling k-core from hornet. What is the motivation for that? 
I though implementations were supposed to be in Hornet as headers and cuGraph would just instantiate and call them. It would be great if the integration pattern was similar to Katz and CoreNumber.

Reply:
Hornet is being used for core_number per vertex call as seen in `gdf_core_number`. This function calculates the relevant subgraph using those results. The subgraph can definitely be calculated in a better way in Hornet however that would require an extra copy and a few other steps. Given that gdf_k_core is already expected to accept filtered vertices as a column, it made more sense to make use of the functions in cugraph.

Reply:
I think that having hornet-based implementations in hornet would be ideal for long term maintenance. Having said that I agree that the subgraph extraction from cugraph is part of the k-core feature and can make things a little bit more challenging. I noticed thrust calls before and after calling the core function, what would be the trade-off if we were to wrap just these? 

Also, can we add a few comments in `gdf_k_core `?